src/transformStateWithClones.js

```diff
@@ -1,13 +1,33 @@
 'use strict';
 
-/**
- * @param {Object} state
- * @param {Object[]} actions
- *
- * @return {Object[]}
- */
-function transformStateWithClones(state, actions) {
-  // write code here
+function transformStateWithClones(initialState, actions) {
+  const states = [];
+
+  let currentState = { ...initialState };
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'clear':
+        currentState = {};
+        break;
+
+      case 'addProperties':
+        currentState = { ...currentState, ...action.extraData };
+        break;
+
+      case 'removeProperties':
+        currentState = { ...currentState };
+
+        for (const key of action.keysToRemove) {
+          delete currentState[key];
+        }
+        break;
+    }
+
+    states.push(currentState);
```

Comment:
Currently, you are pushing the `currentState` object directly into the `states` array. This means all elements in `states` will reference the same object, leading to unexpected results. To fix this, you should push a clone of `currentState` instead. You can achieve this by using `states.push({ ...currentState });` to ensure each state in the array is a separate object.